### ISSUE TITLE
confusing error message for sldb-source-eval-eval-last-expression

### ISSUE THREAD

**Issue by bo-tato** [mmontone/slime-breakpoints#5]
`sldb-source-eval-eval-last-expression` is a really cool feature I feel I'll be using a lot in the future, but it confused me for a bit figuring out how to use it. On breakpoint in debugger and trying to use `sldb-source-eval-eval-last-expression` in my source buffer, I'd get the backtrace:
```
The variable swank/sbcl::*sldb-stack-top* is unbound.
   [Condition of type unbound-variable]

Restarts:
 0: [continue] Retry using swank/sbcl::*sldb-stack-top*.
 1: [use-value] Use specified value.
 2: [store-value] Set specified value and use it.
 3: [*abort] Return to SLIME's top level.
 4: [abort] abort thread (#<thread tid=445 "worker" running {100700FB03}>)

Backtrace:
  0: (swank/sbcl::nth-frame 0)
  1: ((flet swank/backend:eval-in-frame :in "/home/user/.config/emacs/.local/straight/repos/slime/swank/sbcl.lisp") (let ((swank::*sldb-level* 0)) w) 0)
  2: (swank/backend:eval-in-frame (let ((swank::*sldb-level* 0)) w) 0)
  3: (swank::eval-in-frame-aux 0 "w" "CIEL-USER" #<function swank::format-values-for-echo-area>)
  4: (swank:eval-string-in-frame "w" 0 "CIEL-USER")
  5: (sb-int:simple-eval-in-lexenv (swank:eval-string-in-frame "w" 0 "CIEL-USER") #<NULL-LEXENV>)
  6: (eval (swank:eval-string-in-frame "w" 0 "CIEL-USER"))
  7: (swank:eval-for-emacs (swank:eval-string-in-frame "w" 0 "CIEL-USER") "CIEL-USER" 100)
  8: ((lambda nil :in swank::spawn-worker-thread))
  9: (swank/sbcl::call-with-break-hook #<function swank:swank-debugger-hook> #<function (lambda nil :in swank::spawn-worker-thread) {B8024B181B}>)
 10: ((flet swank/backend:call-with-debugger-hook :in "/home/user/.config/emacs/.local/straight/repos/slime/swank/sbcl.lisp") #<function swank:swank-debugger-hook> #<function (lambda nil :in swank::spawn-w..
 11: (swank/backend:call-with-debugger-hook #<function swank:swank-debugger-hook> #<function (lambda nil :in swank::spawn-worker-thread) {B8024B181B}>)
 12: ((lambda nil :in swank::spawn-worker-thread))
 13: (swank::call-with-bindings ((*standard-input* . #<swank/gray::slime-input-stream {1005D084F3}>)) #<function (lambda nil :in swank::spawn-worker-thread) {B8024B1EBB}>)
 14: ((lambda nil :in swank::spawn-worker-thread))
```

I finally figured that in the debugger you need to select goto source on the frame you want to eval in, and then on the buffer that comes up you can use `sldb-source-eval-eval-last-expression`. It seems goto source sets `slime-current-thread`, otherwise it will eval in a different thread where that `*sldb-stack-top*` dynamic variable isn't set and it errors out. I'm not sure the best way to address, maybe two things?
1) In SLDB-SOURCE-EVAL section of readme, mention the need to select show source on the frame you want to eval in
2) in `sldb-source-eval-eval-last-expression` check that `slime-current-thread` is set to some thread number and that `sldb-source-eval-frame-number` is set, otherwise display a nicer error message mentioning the need to select show source on the frame you want to eval in

**Comment by mmontone** [comment]
You are correct, and those are good points. To be sincere I don't think the sldb-source-eval implementation is that great. I did what I could to implement it, but had to deal with the constraints of SLIME implementation, etc. I'll consider your points and try to implement. Thanks.

**Comment by mmontone** [comment]
I've implemented your suggestions. Thanks!

**Comment by bo-tato** [comment]
Thanks! Out of curiosity as I have an interest in development tools and will probably contribute to one or more of slime/sly/lem/neomacs when I have time, what more did you want to do with sldb-source-eval and what constraints of slime did you run into?

**Comment by mmontone** [comment]
I think I found more constraints in the other extension, sldb-show-frame-local. For example, for the highlighting of the frame-locals, I found no model-view-controller pattern, so I had to manipulate the emacs buffer directly (the view), without a model. Things like that ... but it is understandable, a system or feature is designed and implemented to solve a particular problem, and I'm having to twist it to make it do something else.
